cirq/_compat.py

```diff
@@ -83,3 +85,64 @@ def decorated_func(*args, **kwargs) -> Any:
         return decorated_func
 
     return decorator
+
+
+def deprecated_parameter(
+        *,
+        deadline: str,
+        fix: str,
+        func_name: Optional[str] = None,
+        parameter_desc: str,
+        match: Callable[[Tuple[Any, ...], Dict[str, Any]], bool],
+        rewrite: Optional[
+            Callable[[Tuple[Any, ...], Dict[str, Any]],
+                     Tuple[Tuple[Any, ...], Dict[str, Any]]]] = None,
+) -> Callable[[Callable], Callable]:
+    """Marks a function parameter as deprecated.
+
+    Also handles rewriting the deprecated parameter into the new signature.
+
+    Args:
+        deadline: The version where the parameter will be deleted (e.g. "v0.7").
+        fix: A complete sentence describing what the user should be using
+            instead of this particular function (e.g. "Use cos instead.")
+        func_name: How to refer to the function.
+            Defaults to `func.__qualname__`.
+        parameter_desc: The name and type of the parameter being deprecated,
+            e.g. "janky_count" or "janky_count keyword" or
+            "positional janky_count".
+        match: Detects if the deprecated parameter is present in the
+            args/kwargs.
+        rewrite: Returns new args/kwargs that don't use the deprecated
+            parameter. Defaults to making no changes.
+
+    Returns:
+        A decorator that decorates functions with a parameter deprecation
+            warning.
+    """
+
+    def decorator(func: Callable) -> Callable:
+        used = False
+
+        @functools.wraps(func)
+        def decorated_func(*args, **kwargs) -> Any:
+            nonlocal used
+            if match(args, kwargs):
+                if rewrite is not None:
+                    args, kwargs = rewrite(args, kwargs)
+
+                if not used:
+                    used = True
+                    qualname = (func.__qualname__
+                                if func_name is None else func_name)
+                    logging.warning(
```

Comment:
Why `logging.warning` and not `warnings.warn`? If I understand correctly, this will almost never be printed unless the user's script calls `logging.basicConfig(level=logging.WARNING)`

Reply:
`warning.warn("text", DeprecationWarning)` is (sometimes) correctly parsed by pycharm and shows the deprecated function with a strikethrough font and includes the message. I don't know if it's possible to trigger this behavior for deprecated parameters (I know it works for functions) but that would be really nice

Reply:
Seems to work by default for me.

```
$ python -c "import logging; logging.warning('TEST')"
WARNING:root:TEST
```

Also works in the python and ipython repls by default.

Reply:
ok so it does show up, but pycharm doesn't cross it out. I think `warnings.warn("message", DeprecationWarning)` is more common. For example, `logging.warn` uses this form :)

Reply:
For me, `python -c "import warnings; warnings.warn('x', DeprecationWarning)"` is the one that produces no output.

Reply:
Won't that form result in a warning every time you call the method, instead of only the first time?

```
    def warn(self, msg, *args, **kwargs):
        warnings.warn("The 'warn' method is deprecated, "
            "use 'warning' instead", DeprecationWarning, 2)
        self.warning(msg, *args, **kwargs)
```

Reply:
When I use `warnings.warn` in the function, pycharm doesn't start crossing out deprecated functions. I think there's too much separation between the usage inside of the decorator and the function being deprecated by the decorator.

Reply:
I'm going to leave it as is for now. Feel free to do a follow-up that uses `warnings.warn` instead, if you can get pycharm to underline the deprecated methods. That would be really useful.

Reply:
I'm no expert, nor do I really want to become one; but `DeprecationWarning`s get printed only once and there's some complicated behavior for when they get turned on or off. IIRC there's some heuristic that tries to guess if you're running the code as a "developer" or "user"  and it only shows it to users. I'm fine with logging. I don't think pycharm is sophisticated enough to figure out deprecations from decorators